### PR TITLE
feat(deployment):prototype deployment framework

### PR DESCRIPTION
**Pull request by vividf**
## Summary
This PR introduces a new AWML deployment framework that supports deploying and evaluating models using ONNX Runtime and TensorRT.

**See [`deployment/README.md`](https://github.com/vividf/AWML/tree/feat/prototype_deployment_framework/deployment) for details.**

## AWML Deployment Framework

A unified, task-agnostic deployment framework for exporting PyTorch models to ONNX and TensorRT, with comprehensive verification and evaluation capabilities.

#### 1. Unified Deployment Workflow
- Base runner: `BaseDeploymentRunner` + thin project-specific runners (e.g. `CenterPointDeploymentRunner`, `YOLOXDeploymentRunner`, `CalibrationDeploymentRunner`)
- Standardized workflow: **Load Model → Export ONNX → Export TensorRT → Verify → Evaluate**
- Exporters (ONNX / TensorRT) and model wrappers are injected into runners (no hidden/global construction)
- Support for multiple export modes: onnx, trt, both, or none

#### 2. Pipeline Architecture
- **Shared preprocessing/postprocessing** across backends
- **Backend-specific inference** (PyTorch, ONNX, TensorRT)
- Task-specific pipelines: `CenterPointPipeline`
- Consistent interface: `preprocess() → run_model() → postprocess()`

#### 3. Multi-Backend Support
- **PyTorch**: Direct model inference
- **ONNX**: ONNX Runtime inference with model wrapping support
- **TensorRT**: TensorRT engine inference with precision policies (`FP32`, `FP16`, `STRONGLY_TYPED`, `AUTO`)

#### 4. Verification & Evaluation
- **Scenario-based verification**: Compare outputs between backends with configurable tolerance
- **Task-specific evaluation**: mAP for detection, accuracy for classification
- **Performance metrics**: Latency statistics across backends

#### 5. Configuration-Driven
- All settings controlled via deploy config files (`deploy_config.py`)
- Typed exporter configs in exporters/base/configs.py: 
  - `ONNXExportConfig`, `TensorRTExportConfig`
  - `TensorRTModelInputConfig`, `TensorRTProfileConfig`
- Configs can be created from dictionaries (bridge between config files and type-safe code)
- Project-specific customizations supported via project runners, exporters, wrappers, data loaders, and evaluators

### Required Files per Project
Each project provides its own deployment entry point, data loader, evaluator, and config:
```
projects/<ProjectName>/deploy/
├── main.py                    # Entry point script
├── evaluator.py               # Project-specific evaluator (inherits BaseEvaluator)
├── data_loader.py             # Project-specific data loader (inherits BaseDataLoader)
└── configs/
    └── deploy_config.py       # Deployment configuration
```
Exporters and runners live in the shared deployment/ package and are reused across projects

### Usage Example

```bash
# Deploy a model
python projects/X/deploy/main.py \
    configs/deploy_config.py \
    configs/model_config.py \
```

### CODE REVIEW

File: deployment/core/base_config.py [hunk 1]
```diff
@@ -0,0 +1,522 @@
...
+class BackendConfig:
+    """Configuration for backend-specific settings."""
+
+    common_config: Mapping[str, Any] = field(default_factory=_empty_mapping)
+    model_inputs: Tuple[TensorRTModelInputConfig, ...] = field(default_factory=tuple)
+
+    @classmethod
+    def from_dict(cls, config_dict: Dict[str, Any]) -> "BackendConfig":
+        common_config = dict(config_dict.get("common_config", {}))
+        model_inputs_raw: Iterable[Dict[str, Any]] = config_dict.get("model_inputs", []) or []
```
Comment (KSeangTan) [inline deployment/core/base_config.py]: why `Iterable`?

File: deployment/README.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
+# AWML Deployment Framework
+
+AWML ships a unified, task-agnostic deployment stack that turns trained PyTorch
+checkpoints into production-ready ONNX and TensorRT artifacts. The same typed
```
Comment (KSeangTan) [inline deployment/README.md]: `The type verification`?

File: deployment/core/io/base_data_loader.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        ground_truth: Labels or annotations if available.
+        metadata: Additional information required for evaluation.
+    """
+
+    input: Any
+    ground_truth: Any
+    metadata: Dict[str, Any]
+
+
+class BaseDataLoader(ABC):
```
Comment (KSeangTan) [inline deployment/core/io/base_data_loader.py]: Any reason why we need this `BaseDataLoader`, but not using the default `dataloader` from mmdetection first?

File: deployment/core/io/preprocessing_builder.py [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+def _build_pipeline(task_type: str, pipeline_cfg: List[TransformConfig]) -> Any:
+    """Build pipeline for a given task_type using registered builders."""
+    try:
+        builder = _PIPELINE_BUILDERS[task_type]
+    except KeyError:
+        raise ValueError(f"Unknown task_type '{task_type}'. " f"Must be one of {VALID_TASK_TYPES}")
+    return builder(pipeline_cfg)
+
+
+def build_preprocessing_pipeline(
```
Comment (KSeangTan) [inline deployment/core/io/preprocessing_builder.py]: We should make `task_type` mandatory instead of optional, in this case, we don't need to `_resolve_task_type` separately, and we only need to check:

```
if task_type not in VALID_TASK_TYPES
```

File: deployment/core/metrics/classification_metrics.py [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+    Attributes:
+        class_names: List of class names for evaluation (e.g., ["miscalibrated", "calibrated"]).
+        frame_id: Frame ID for evaluation (not used for classification but kept for consistency).
+    """
+
+    # Override default frame_id for classification (not actually used but kept for interface consistency)
+    frame_id: str = "classification"
+
+
+class ClassificationMetricsAdapter(BaseMetricsAdapter):
```
Comment (KSeangTan) [inline deployment/core/metrics/classification_metrics.py]: Ideally, it should just an interface to call a metric framework? So the internal metric computation should be in another place that can be used by all classification projects?

File: deployment/runners/deployment_runner.py [hunk 6]
```diff
@@ -0,0 +1,859 @@
...
+
+from deployment.core import Artifact, Backend, BaseDataLoader, BaseDeploymentConfig, BaseEvaluator, ModelSpec
+from deployment.exporters.base.factory import ExporterFactory
+from deployment.exporters.base.model_wrappers import BaseModelWrapper
+from deployment.exporters.base.onnx_exporter import ONNXExporter
+from deployment.exporters.base.tensorrt_exporter import TensorRTExporter
+from deployment.exporters.workflows.base import OnnxExportWorkflow, TensorRTExportWorkflow
+
+
+class DeploymentResultDict(TypedDict, total=False):
```
[the remark this answers is not in the document]
Comment (vividf) [reply deployment/runners/deployment_runner.py]: Thanks, addressed in https://github.com/tier4/AWML/pull/137/commits/480a2aab262546adb6c1710d274cc9f3ed0690f6

File: deployment/core/config/base_config.py [hunk 7]
```diff
@@ -0,0 +1,522 @@
...
+    @classmethod
+    def from_dict(cls, config_dict: Dict[str, Any]) -> "ExportConfig":
+        """Create ExportConfig from dict."""
+        return cls(
+            mode=ExportMode.from_value(config_dict.get("mode", ExportMode.BOTH)),
+            work_dir=config_dict.get("work_dir", cls.work_dir),
+            checkpoint_path=config_dict.get("checkpoint_path"),
+            onnx_path=config_dict.get("onnx_path"),
+            tensorrt_path=config_dict.get("tensorrt_path"),
+            cuda_device=config_dict.get("cuda_device", cls.cuda_device),
```
[the remark this answers is not in the document]
Comment (vividf) [reply deployment/core/config/base_config.py]: This was outdated in the  previous comment, would you mind checking again

File: deployment/README.md [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+verification and evaluation toolchain runs across every backend so numerical
+parity and metrics stay consistent from project to project.
+
+At the center is a shared runner/pipeline/exporter architecture that teams can
+extend with lightweight wrappers or workflows. CenterPoint, YOLOX,
+CalibrationStatusClassification, and future models plug into the same export and
```
[the remark this answers is not in the document]
Comment (vividf) [reply deployment/README.md]: It's export workflow

File: deployment/core/evaluation/base_evaluator.py [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+            if verbose and idx % EVALUATION_DEFAULTS.LOG_INTERVAL == 0:
+                logger.info(f"Processing sample {idx + 1}/{actual_samples}")
+
+            sample = data_loader.load_sample(idx)
+            input_data, infer_kwargs = self._prepare_input(sample, data_loader, model.device)
+
+            gt_data = data_loader.get_ground_truth(idx)
+            ground_truths = self._parse_ground_truths(gt_data)
+
+            raw_output, latency, breakdown = pipeline.infer(input_data, **infer_kwargs)
```
[the remark this answers is not in the document]
Comment (vividf) [reply deployment/core/evaluation/base_evaluator.py]: Agree, thanks, addressed in https://github.com/tier4/AWML/pull/137/commits/142fc0d10e05a0048ceb724b8f93a7235998b898

File: deployment/core/io/preprocessing_builder.py [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+    "detection3d": _build_detection3d,
+    "classification": _build_classification,
+    "segmentation": _build_segmentation,
+}
+
+# Valid task types
+VALID_TASK_TYPES = list(_PIPELINE_BUILDERS.keys())
+
+
+def _build_pipeline(task_type: str, pipeline_cfg: List[TransformConfig]) -> Any:
```
[the remark this answers is not in the document]
Comment (vividf) [reply deployment/core/io/preprocessing_builder.py]: Thanks, addressed in https://github.com/tier4/AWML/commit/28446906f34a5bfc8bded99a2f384c6c99a70db9

File: deployment/runners/common/export_orchestrator.py [hunk 11]
```diff
@@ -0,0 +1,520 @@
...
+                if pytorch_model is None:
+                    return result
+                result.pytorch_model = pytorch_model
+
+            onnx_artifact = self._export_onnx(pytorch_model, context)
+            if onnx_artifact:
+                result.onnx_path = onnx_artifact.path
+
+        # Step 4: Export TensorRT if requested
+        if should_export_trt:
```
Comment (KSeangTan) [inline deployment/runners/common/export_orchestrator.py]: Please move this part to another class function

**Review by KSeangTan — APPROVED** [review]
LGTM overall
